CI: add Win 3.10

tylerjereddy opened the pull request:
Deals with the Windows portion of gh-3457

Changes made in this Pull Request:
 - adds Windows Python 3.10 testing to Azure CI to prevent any regressions there moving forward
 - for the moment, we skip `h5py` install on Python 3.10 and we no longer pin the version of `gsd`
 - we temporarily suppress the following Python 3.10 warning observed on Windows for a third-party library that is still importing `six`: `ImportWarning('_SixMetaPathImporter.find_spec() not found`
 - add the option to skip one test that had an obligate requirement for `h5py`, which seems more consistent anyway because we already have a bunch of tests that skip conditionally with `HAS_H5PY`
 - in my hands, this allows the test suite to pass on Windows w/ Python `3.10`; I do suggest that we open an issue to revert any of the minor workarounds in use here as the `3.10` dependencies/ecosystem matures in the future


PR Checklist
------------
 - [ ] Tests?
 - [ ] Docs?
 - [ ] CHANGELOG updated?
 - [ ] Issue raised/referenced?

tylerjereddy committed:
CI: add Win 3.10

pep8speaks commented:
Hello @tylerjereddy! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`testsuite/MDAnalysisTests/coordinates/test_chainreader.py`](https://github.com/MDAnalysis/mdanalysis/blob/ba91144e866d8bf44a733271e995aa7627759260/testsuite/MDAnalysisTests/coordinates/test_chainreader.py):

> [Line 357:17](https://github.com/MDAnalysis/mdanalysis/blob/ba91144e866d8bf44a733271e995aa7627759260/testsuite/MDAnalysisTests/coordinates/test_chainreader.py#L357): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent



* In the file [`testsuite/MDAnalysisTests/topology/test_top.py`](https://github.com/MDAnalysis/mdanalysis/blob/ba91144e866d8bf44a733271e995aa7627759260/testsuite/MDAnalysisTests/topology/test_top.py):

> [Line 502:17](https://github.com/MDAnalysis/mdanalysis/blob/ba91144e866d8bf44a733271e995aa7627759260/testsuite/MDAnalysisTests/topology/test_top.py#L502): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent





##### Comment last updated at 2021-12-26 21:54:56 UTC

tylerjereddy commented:
The codecov failure looks spurious (we probably should never fail CI on codecov anyway, there can be various reasons not to want that).

The PEP8 complaint is probably debatable--feel free to push in an adjustment for that if you don't like the visual indents or whatever.

IAlibay commented:
@tylerjereddy is the py3.10 h5py skip a known issue?

tylerjereddy commented:
> @tylerjereddy is the py3.10 h5py skip a known issue?

@IAlibay At the time I opened this PR there were no Windows Python 3.10 binary wheels up on PyPI. I believe there now are, so you could just try removing that, or I can do that in the next few days if you want.

tylerjereddy commented on azure-pipelines.yml:
```diff
@@ -54,17 +57,20 @@ jobs:
+      python -m pip install --only-binary=h5py
+      "h5py<3.4.0"
+    displayName: 'Install h5py for Python < 3.10'
+    condition: and(succeeded(), ne(variables['PYTHON_VERSION'], '3.10'))
```
we could try removing this condition, though we'll need to unpin `h5py` to get the binary on newer Python, and not sure how that will work out..

tylerjereddy committed:
MAINT: PR 3461 revisions

* use `h5py` for all Python versions, including
`3.10`

* restore the usage of the `gsd` package; it was
only meant to be unpinned, not deleted

* deal with PEP8 complaints

tylerjereddy commented:
@IAlibay I've simplified the diff a bit here--I'm almost tempted to `xfail` the single Windows h5py-related test failures for now. Some kind of permissions issue.

We also need to get on top of the Windows version bump in CI--I think we have other issues/PRs for that, but version 2016 is going away soon I think.

IAlibay commented on testsuite/MDAnalysisTests/coordinates/test_chainreader.py:
```diff
@@ -345,7 +348,17 @@ def test_easy_trigger_warning(self, tmpdir):
+            with warnings.catch_warnings():
+                # for windows Python 3.10 ignore:
+                # ImportWarning('_SixMetaPathImporter.find_spec() not found
+                # TODO: remove when we no longer have a dependency
```
Seems sensible to me.

Could this be raised as an issue so we remember to deal with this in the future?

tylerjereddy replied:
see gh-3485 for the reminder; also, note that the `xfail` has been restored from previously being commented out--so this test has clearly been a pain point in the past (we could also be more granular with the `xfail`, since it only occurs for one of the parameters I think, but since the `xfail` was previously used more broadly I've simply restored it for Windows).

IAlibay reviewed (changes requested):
Only putting this as a "request changes" with the ask that we raise some issues + the h5py xfail.

Regarding h5py - I don't think this is something we can fix immediately (it looks like something we'll need to deal with upstream), so I agree with the xfail. Just if we can also raise an issue to remind us to deal with it, that'd be awesome.

tylerjereddy committed:
MAINT: PR 3461 revisions

* restore the `xfail` for `test_write_with_drivers`
on Windows for now, as this test seems problematic
once again for newer versions of `h5py`

IAlibay reviewed (approved):
Thanks @tylerjereddy !
